src/components/Modal/Modal.tsx

```diff
@@ -1,83 +1,110 @@
-import React, { PureComponent } from "react";
+import React, { PureComponent, ReactNode } from "react";
 import { createPortal } from "react-dom";
-import PropTypes from "prop-types";
 import cn from "classnames";
 import Button from "~/components/Button";
 import ModalContext from "./Modal.context";
 import { timeout } from "./constants";
 import "./Modal.css";
 
-let _root;
+let _root: any;
 
-class Modal extends PureComponent {
+type Props = {
+  children: ReactNode;
+  isOpen: boolean;
+  onClose: () => void;
+  fullScreen?: boolean;
+  className: string;
+}
+
+type State = {
+  isMouseClicked: boolean;
+  isAboutToClose: boolean;
+  isClosedGracefully: boolean;
+}
+
+class Modal extends PureComponent<Props, State> {
   static Context = ModalContext;
   static timeout = timeout;
 
-  static propTypes = {
-    children: PropTypes.node,
-    isOpen: PropTypes.bool,
-    onClose: PropTypes.func,
-    fullScreen: PropTypes.bool
-  };
+  state = {
+    isMouseClicked: false,
+    isAboutToClose: false,
```

Comment:
I think this breaks a functionality. Previously clicking outside of the `Modal` would close the modal. Now that functionality is broken - nothing happens when I click outside. I think (though I'm not 100% sure), it's because React batches the `state` updates and therefore `gracefulClose` is not able to react properly. If we revert these to being instance variables I think the functionality should be fixed. 

We can think of these variables as flags, they're not really representing the `State` of the component therefore I think it's safe leveraging instance variables.

Reply:
it's a problem of async state update, I answered below :)